apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        if (challengerFeeToken != feeToken) {
+            _transfer(challengerFeeToken, _challenge.challenger, challengerFeeAmount);
+        }
+
+        return disputeId;
+    }
+
+    /**
+    * @dev Update evidence submission status for a disputed action
+    * @param _action Action instance whose dispute is being submitted evidence
+    * @param _challenge Current challenge instance for the action
+    * @param _submitter Address submitting the evidence
+    * @param _finished Whether the evidence submitter is finished submitting evidence
+    * @returns Whether both parties have finished submitting evidence
+    */
+    function _updateEvidenceSubmissionStatus(Action storage _action, Challenge storage _challenge, address _submitter, bool _finished)
+        internal
+        returns (bool)
+    {
+        bool submitterFinishedEvidence = _challenge.submitterFinishedEvidence;
+        bool challengerFinishedEvidence = _challenge.challengerFinishedEvidence;
+
+        if (_submitter == _action.submitter) {
+            require(!submitterFinishedEvidence, ERROR_SUBMITTER_FINISHED_EVIDENCE);
+            if (_finished) {
+                submitterFinishedEvidence = _finished;
+                _challenge.submitterFinishedEvidence = _finished;
+            }
+        } else if (_submitter == _challenge.challenger) {
+            require(!challengerFinishedEvidence, ERROR_CHALLENGER_FINISHED_EVIDENCE);
+            if (_finished) {
+                submitterFinishedEvidence = _finished;
+                _challenge.challengerFinishedEvidence = _finished;
+            }
+        } else {
+            revert(ERROR_SENDER_NOT_ALLOWED);
+        }
+
+        return submitterFinishedEvidence && challengerFinishedEvidence;
+    }
+
+    /**
+    * @dev Submit evidence for an dispute on an arbitrator
+    * @param _arbitrator Arbitrator to submit evidence on
+    * @param _disputeId Identification number of the dispute on the arbitrator
+    * @param _submitter Address submitting the evidence
+    * @param _evidence Evidence data submitted for the dispute
+    * @param _finished Whether the submitter is finished submitting evidence
+    */
+    function _submitEvidence(IArbitrator _arbitrator, uint256 _disputeId, address _submitter, bytes _evidence, bool _finished) internal {
+        if (_evidence.length > 0) {
+            emit EvidenceSubmitted(_arbitrator, _disputeId, _submitter, _evidence, _finished);
+        }
+    }
+
+    /**
+    * @dev Accept a challenge proposed against an action ("reject action")
+    * @param _actionId Identification number of the action to be rejected
+    * @param _action Action instance to be rejected
+    * @param _challengeId Current challenge identification number for the action
+    * @param _challenge Current challenge instance for the action
+    */
+    function _acceptChallenge(uint256 _actionId, Action storage _action, uint256 _challengeId, Challenge storage _challenge) internal {
```

Comment:
Wondering if we can find a better name for these three functions.

They make sense, but I find them to be confusing at a glance because we use both "challenge accepted" and "action rejected" as concepts that are synonymous with each other.

Reply:
Yea, I agree. For the moment I changed them to be in terms of the action instead, as that is what we are doing for the events and callbacks